File: Python/11. 서버저장조건 조건검색.py
import asyncio
import ebest
from common import *
from app_keys import *

async def main():
    api=ebest.OpenApi()
    if not await api.login(appkey, appsecretkey): return print(f'연결실패: {api.last_message}')
    
    # 조건검색식 리스트 조회
    request = {
        't1866InBlock': {
            'user_id': user_id, # 사용자ID 8자리
            'gb': '0', # 0 : 그룹+조건리스트 조회, 1 : 그룹리스트조회, 2 : 그룹명에 속한 조건리스트조회
            'group_name': '',
            'cont': '0',
            'cont_key': '',
        }
    }
    response = await api.request('t1866', request)
    if not response: return print(f'요청실패: {api.last_message}')
    
    cond_list = response.body['t1866OutBlock1']
    print_table(cond_list)
    
    # 요청할 조건검색식 선택
    cond_len = len(cond_list)
    sel_index = int(input(f'조건검색식index (0~{cond_len-1})를 입력하세요:'))
    if sel_index >= cond_len: return print('잘못된 index')
    
    # 조건검색식 조회
    query_index = cond_list[sel_index]['query_index']
    request = {
        't1859InBlock': {
            'query_index': query_index,
        }
    }
    response = await api.request('t1859', request)
    if not response: return print(f'요청실패: {api.last_message}')
    
    item_list = response.body.get('t1859OutBlock1', None)
    if item_list:
        print_table(item_list)
    else:
        print(f'조건검색식[{query_index}] 결과 없음')
    
    ... # 다른 작업 수행
    await api.close()


asyncio.run(main())

# Output:
'''
Row Count = 1
+--------------+------------+------------+
| query_index  | group_name | query_name |
+--------------+------------+------------+
| XXXXXXXX0001 |  조건전략  |  고가돌파  |
+--------------+------------+------------+
조건검색식index (0~0)를 입력하세요:0
Row Count = 20
+--------+-----------------------------------+-------+------+--------+-------+--------+
| shcode |               hname               | price | sign | change |  diff | volume |
+--------+-----------------------------------+-------+------+--------+-------+--------+
| 005670 |               푸드웰              |  4770 |  2   |   40   |  0.85 |  5964  |
| 009580 |              무림P&P              |  3105 |  5   |   15   | -0.48 | 44058  |
| 009680 |               모토닉              |  8260 |  5   |   10   | -0.12 | 18460  |
| 028100 |              동아지질             | 13750 |  5   |  100   | -0.72 | 21003  |
| 032560 |             황금에스티            |  7050 |  5   |   50   | -0.70 | 40532  |
| 032750 |                삼진               |  5400 |  5   |   10   | -0.18 | 18638  |
...
'''

File: CSharp/11. 서버저장조건 조건검색.cs
namespace CSharp;

internal class _11 : SampleBase
{
    public override async Task ActionImplement()
    {
        // [요청] t1866 : 서버저장조건리스트조회(API)
        t1866 tr_data = new()
        {
            t1866InBlock = new(Secret.UserId, "0", "", "0", ""),
        };
        await api.GetTRData(tr_data);
        if (tr_data.t1866OutBlock is null)
        {
            // 오류 처리
            print(tr_data.rsp_cd.Length > 0 ? $"{tr_data.rsp_cd}-{tr_data.rsp_msg}" : api.LastErrorMessage);
            return;
        }

        // tr_data.t1866OutBlock 데이터 처리
        print(tr_data.t1866OutBlock);
        print(tr_data.t1866OutBlock1);

        if (tr_data.t1866OutBlock1 == null)
        {
            print("서버에 저장된 조건검색식이 없습니다.");
            return;
        }

        // 요청할 조건검색식 선택
        int sel_index;
        do
        {
            Console.Write($"조건검색식index (0~{tr_data.t1866OutBlock.result_count - 1})를 입력하세요:");
            int.TryParse(Console.ReadLine(), out sel_index);
            if (sel_index < 0 || sel_index >= tr_data.t1866OutBlock.result_count)
            {
                print("잘못된 입력입니다.");
                sel_index = -1;
            }
        } while (sel_index == -1);

        // [요청] t1859 : 서버저장조건 조건검색
        t1859 tr_data_cond = new()
        {
            t1859InBlock = new(tr_data.t1866OutBlock1[sel_index].query_index),
        };
        await api.GetTRData(tr_data_cond);
        if (tr_data_cond.t1859OutBlock is null)
        {
            // 오류 처리
            print(tr_data_cond.rsp_cd.Length > 0 ? $"{tr_data_cond.rsp_cd}-{tr_data_cond.rsp_msg}" : api.LastErrorMessage);
            return;
        }

        // tr_data.t1859OutBlock 데이터 처리
        print(tr_data_cond.t1859OutBlock);
        print(tr_data_cond.t1859OutBlock1);
    }
}

// Output:
/*
t1866OutBlock, Field Count = 3
| Key          | Value |
|--------------|-------|
| result_count | 1     |
| cont         |       |
| contkey      |       |

t1866OutBlock1[], Field Count = 3, Data Count = 1
| query_index  | group_name | query_name |
|--------------|------------|------------|
| XXXXXXXX0001 | 나의전략   | 조건전략   |

조건검색식index (0~0)를 입력하세요:0
t1859OutBlock, Field Count = 3
| Key          | Value  |
|--------------|--------|
| result_count | 4     |
| result_time  | 142055 |
| text         |        |

t1859OutBlock1[], Field Count = 7, Data Count = 34
| shcode | hname                           | price | sign | change |  diff |  volume |
|--------|---------------------------------|-------|------|--------|-------|---------|
| 002390 | 한독                            | 14510 | 2    |    710 |  5.14 |   80702 |
| 002795 | 아모레G우                       | 10360 | 5    |     60 | -0.58 |    2816 |
| 003200 | 일신방직                        |  9090 | 2    |    140 |  1.56 |   46689 |
| 003475 | 유안타증권우                    |  2650 | 5    |     30 | -1.12 |    8770 |
*/